# Lab 6: Làm việc với Dữ liệu trong Python

## Đọc File trong Python

### Download Data

In [1]:
import urllib.request

url = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%204/data/example1.txt'
filename = 'Example1.txt'
urllib.request.urlretrieve(url, filename)

('Example1.txt', <http.client.HTTPMessage at 0x7fc7b2a38d50>)

### Reading text file

In [2]:
example1 = 'Example1.txt'
file1 = open(example1, 'r')
file222 = open(example1)
type(file222)
file222.read()

'This is line 1 \nThis is line 2\nThis is line 3'

In [3]:
file1.name

'Example1.txt'

In [4]:
file1.mode

'r'

In [5]:
FileContent = file1.read()
FileContent

'This is line 1 \nThis is line 2\nThis is line 3'

In [6]:
print(FileContent)

This is line 1 
This is line 2
This is line 3


In [7]:
file1.close()

### Better way to read text file

Cách tốt hơn là sử dụng câu lệnh <code>with</code>, nó sẽ tự động đóng file ngay cả khi code gặp phải một ngoại lệ. Code sẽ chạy mọi thứ trong khối thụt lề sau đó đóng đối file object.

In [8]:
with open(example1, 'r') as file1:
    FileContent = file1.read()
    print(FileContent)

This is line 1 
This is line 2
This is line 3


Ví dụ: chúng ta không phải đọc toàn bộ file, có thể đọc 4 ký tự đầu tiên bằng cách nhập 3 ký tự làm tham số cho phương thức **.read()**:

In [9]:
with open(example1, 'r') as file1:
    FileContent = file1.read(4)
    print(FileContent)

This


Khi phương thức <code>.read(4)</code> được gọi, 4 ký tự đầu tiên được gọi. Nếu chúng ta gọi lại phương thức, 4 ký tự tiếp theo sẽ được gọi. Output cho cell sau sẽ minh họa quy trình cho các input khác nhau cho phương thức <code>read()</code>:

In [10]:
with open(example1, 'r') as file1:
    print(file1.read(4))
    print(file1.read(4))
    print(file1.read(7))
    print(file1.read(15))

This
 is 
line 1 

This is line 2


In [11]:
with open(example1, 'r') as file1:
    print(file1.readline())

This is line 1 



Chúng ta cũng có thể truyền một đối số tới <code> readline()</code> để chỉ định số ký tự mà chúng ta muốn đọc. Tuy nhiên, không giống như <code> read()</code>, <code> readline()</code> chỉ có thể đọc nhiều nhất một dòng.

In [12]:
with open(example1, 'r') as file1:
    print(file1.readline(20)) # readline(20) reads the first 20 characters
   

This is line 1 



In [13]:
with open(example1, 'r') as file1:
    for line in file1:
        print(line)

This is line 1 

This is line 2

This is line 3


In [14]:
with open(example1, 'r') as file1:
    filelist = file1.readlines()
    for line in filelist:
        print(type(line))

print(file1.closed)

filelist[0]

<class 'str'>
<class 'str'>
<class 'str'>
True


'This is line 1 \n'

## Đọc và Lưu file trong Python

In [15]:
example2 = 'Example2.txt'
with open(example2, 'w') as file2:
    file2.write('This is line A')

In [16]:
with open(example2, 'r') as file2:
    print(file2.read())

This is line A


In [17]:
with open(example2, 'w') as writefile:
    writefile.write("This is line A\n")
    writefile.write("This is line B\n")

In [18]:
with open(example2, 'r') as file2:
    print(file2.read())

This is line A
This is line B



### Thêm vào file

In [19]:
with open(example2, 'a') as testwritefile:
    testwritefile.write("This is line C\n")
    testwritefile.write("This is line D\n")
    testwritefile.write("This is line E\n")

### Các mode khác

Việc mở file bằng  **a** hoặc **w** và sau đó mở lại trong **r** để đọc bất kỳ dòng nào khá là không hiệu quả. May mắn thay, chúng ta có thể truy cập file ở các mode sau:

-   **r+**  Đọc và ghi. Không thể cắt ngắn file.
-   **w+** : Ghi và đọc. Cắt bớt file.
-   **a+** : Ghi thêm và Đọc. Tạo một file mới nếu không có file nào tồn tại. Bạn không cần phải tìm hiểu chi tiết cụ thể của từng mode cho lab này. 


In [20]:
with open(example2, 'a+') as testwritefile:
    testwritefile.write("This is line I\n")
    print(testwritefile.read())
    

Không có lỗi nhưng <code>read() </code> cũng không xuất ra bất kỳ thứ gì. Điều này là do vị trí của chúng ta trong file.

Hầu hết các phương thức file mà chúng ta đã xem đều hoạt động ở một vị trí nhất định trong file. <code>.write() </code> ghi tại một vị trí nhất định trong file. <code>.read()</code> đọc tại một vị trí nhất định trong file,... Bạn có thể coi điều này giống như việc di chuyển con trỏ của bạn trong notepad để thực hiện các thay đổi tại vị trí cụ thể.

Mở file trong **w** cũng giống như mở file .txt, di chuyển con trỏ của bạn đến đầu file văn bản, viết văn bản mới và xóa mọi thứ sau đó.
Trong khi đó, việc mở file bằng **a** tương tự như mở file .txt, di chuyển con trỏ của bạn đến cuối và sau đó thêm các đoạn văn bản mới. <br>
Thường rất hữu ích khi biết vị trí của 'con trỏ' trong file và có thể kiểm soát nó. Các phương thức sau đây cho phép chúng ta thực hiện điều này chính xác -

-   <code>.tell()</code> - trả về vị trí hiện tại tính bằng byte
-   <code>.seek(offset,from)</code> - thay đổi vị trí của byte 'offset' đối với 'from'. From có thể lấy giá trị 0,1,2 tương ứng với phần đầu, liên quan đến vị trí hiện tại và phần cuối



In [21]:
with open(example2, 'a+') as file2:
    print('Initial Location {}'.format(file2.tell()))
    data = file2.read()
    if (not data):
        print('Read Nothing ')
    else:
        print(data)
    
    file2.seek(0,0) # seek(0,0) moves the file pointer to the beginning of the file
    print('After seek Location {}'.format(file2.tell()))
    data = file2.read()
    if (not data):
        print('Read Nothing ')
    else:
        print(data)
    print('After read Location {}'.format(file2.tell()))

Initial Location 90
Read Nothing 
After seek Location 0
This is line A
This is line B
This is line C
This is line D
This is line E
This is line I

After read Location 90


Cuối cùng, một lưu ý về sự khác biệt giữa **w+** và **r+**. Cả hai mode này đều cho phép truy cập các phương thức đọc và ghi. Tuy nhiên, việc mở file trong **w+** sẽ ghi đè file đó và xóa tất cả dữ liệu hiện có. <br>
Để làm việc với file trên dữ liệu hiện có, hãy sử dụng **r+** và **a+**. Trong khi sử dụng **r+**, bạn nên thêm phương thức <code>.truncate()</code> vào cuối dữ liệu của bạn. Điều này sẽ giảm file dữ liệu của bạn và xóa mọi thứ đi cùng  phần giảm đó. <br>
Trong khối code sau, chạy code như lúc đầu và sau đó chạy với <code>.truncate()</code>.

In [22]:
with open(example2, 'r+') as file2:
    data = file2.readlines()
    file2.seek(0,0)

    file2.write('This is line A\n')
    file2.write('This is line B\n')
    file2.write('This is line C\n')
    file2.write('Finished\n')

    file2.seek(0,0)
    print(file2.read())


This is line A
This is line B
This is line C
Finished
ine D
This is line E
This is line I



In [23]:
with open(example2, 'r+') as file2:
    data = file2.readlines()
    file2.seek(0,0)

    file2.write('This is line A\n')
    file2.write('This is line B\n')
    file2.write('This is line C\n')
    file2.write('Finished\n')

    file2.truncate()
    file2.seek(0,0)
    print(file2.read())

This is line A
This is line B
This is line C
Finished



### Sao chép file

Hãy sao chép file Example2.txt vào file Example3.txt:

In [24]:
with open('Example2.txt', 'r') as file2:
    with open('Example3.txt', 'w') as file3:
        for line in file2:
            file3.write(line)

### **Exercise**

Câu lạc bộ người hâm mộ Raptors của trường đại học tại địa phương của bạn duy trì danh sách các thành viên tích cực của mình trên tài liệu .txt. Hàng tháng, họ cập nhật file bằng cách loại bỏ các thành viên không hoạt động. Bạn đã được giao nhiệm vụ tự động hóa điều này bằng các kỹ năng python của mình. <br>
Với file currentMem, loại từng thành viên có 'no' trong cột Active. Theo dõi từng thành viên đã bị loại bỏ và thêm họ vào file exMem. Hãy đảm bảo định dạng của file gốc được giữ nguyên.
(_Gợi ý: Làm điều này bằng cách đọc/ghi toàn bộ các dòng và đảm bảo vẫn giữ nguyên tiêu đề_)
<br>
Chạy khối code bên dưới trước khi bắt đầu exercise. Khung code đã cung cấp cho bạn, bạn chỉ chỉnh sửa hàm cleanFiles.

In [25]:
from random import randint as rnd

memReg = 'members.txt'
exReg = 'inactive.txt'
fee =('yes','no')

def genFiles(current,old):
    with open(current,'w+') as writefile: 
        writefile.write('Membership No  Date Joined  Active  \n')
        data = "{:^13}  {:<11}  {:<6}\n"

        for rowno in range(20):
            date = str(rnd(2015,2020))+ '-' + str(rnd(1,12))+'-'+str(rnd(1,25))
            writefile.write(data.format(rnd(10000,99999),date,fee[rnd(0,1)]))


    with open(old,'w+') as writefile: 
        writefile.write('Membership No  Date Joined  Active  \n')
        data = "{:^13}  {:<11}  {:<6}\n"
        for rowno in range(3):
            date = str(rnd(2015,2020))+ '-' + str(rnd(1,12))+'-'+str(rnd(1,25))
            writefile.write(data.format(rnd(10000,99999),date,fee[1]))


genFiles(memReg,exReg)

In [26]:
def cleanFiles(currentMem,exMem):
    '''
    currentMem: File containing list of current members
    exMem: File containing list of old members
    
    Removes all rows from currentMem containing 'no' and appends them to exMem
    '''
    with open(currentMem,'r+') as readfile:
        with open(exMem,'a+') as writefile:
            members = readfile.readlines()
            headers = members[0]
            members.pop(0)
            readfile.seek(0,0)
            readfile.truncate()
            readfile.write(headers)
            for member in members:
                if 'no' in member:
                    writefile.write(member)
                else:
                    readfile.write(member)
            


# Code giúp bạn thấy file
# Để như sau
memReg = 'members.txt'
exReg = 'inactive.txt'
cleanFiles(memReg,exReg)


headers = "Membership No  Date Joined  Active  \n"
with open(memReg,'r') as readFile:
    print("Active Members: \n\n")
    print(readFile.read())
    
with open(exReg,'r') as readFile:
    print("Inactive Members: \n\n")
    print(readFile.read())

Active Members: 


Membership No  Date Joined  Active  
    72629      2016-7-11    yes   
    55061      2016-7-15    yes   
    47447      2016-1-11    yes   
    48147      2019-6-24    yes   
    96796      2019-9-21    yes   
    23884      2019-10-10   yes   
    97622      2019-4-1     yes   
    89724      2019-12-12   yes   

Inactive Members: 


Membership No  Date Joined  Active  
    84318      2020-6-12    no    
    44060      2018-3-22    no    
    95765      2015-4-24    no    
    82389      2015-1-3     no    
    90949      2015-4-22    no    
    61369      2019-8-17    no    
    75820      2019-6-2     no    
    48249      2015-5-15    no    
    68724      2019-7-22    no    
    97783      2018-10-16   no    
    16268      2017-11-25   no    
    27227      2020-12-3    no    
    26385      2016-9-22    no    
    56225      2016-3-3     no    
    35416      2020-6-3     no    



In [27]:
def testMsg(passed):
    if passed:
       return 'Test Passed'
    else :
       return 'Test Failed'

testWrite = "testWrite.txt"
testAppend = "testAppend.txt" 
passed = True

genFiles(testWrite,testAppend)

with open(testWrite,'r') as file:
    ogWrite = file.readlines()

with open(testAppend,'r') as file:
    ogAppend = file.readlines()

try:
    cleanFiles(testWrite,testAppend)
except:
    print('Error')

with open(testWrite,'r') as file:
    clWrite = file.readlines()

with open(testAppend,'r') as file:
    clAppend = file.readlines()
        
# kiểm tra xem không có bất kỳ hàng nào giống nhau, bao gồm tiêu đề

if (len(ogWrite) + len(ogAppend) != len(clWrite) + len(clAppend)):
    print("The number of rows do not add up. Make sure your final files have the same header and format.")
    passed = False
    
for line in clWrite:
    if  'no' in line:
        passed = False
        print("Inactive members in file")
        break
    else:
        if line not in ogWrite:
            print("Data in file does not match original file")
            passed = False
print ("{}".format(testMsg(passed)))

Test Passed


## Giới thiệu về Python Pandas

In [28]:
# Dependency cần để cài file 

%pip install xlrd

Note: you may need to restart the kernel to use updated packages.


In [29]:
import pandas as pd

In [30]:
csv_path = 'https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%204/Datasets/TopSellingAlbums.csv'
df = pd.read_csv(csv_path)

In [31]:
df.head()

Artist                      Album  Released   Length  \
0  Michael Jackson                   Thriller      1982  0:42:19   
1            AC/DC              Back in Black      1980  0:42:11   
2       Pink Floyd  The Dark Side of the Moon      1973  0:42:49   
3  Whitney Houston              The Bodyguard      1992  0:57:44   
4        Meat Loaf            Bat Out of Hell      1977  0:46:33   

                         Genre  Music Recording Sales (millions)  \
0               pop, rock, R&B                              46.0   
1                    hard rock                              26.1   
2             progressive rock                              24.2   
3               R&B, soul, pop                              27.4   
4  hard rock, progressive rock                              20.6   

   Claimed Sales (millions) Released.1 Soundtrack  Rating  
0                        65  30-Nov-82        NaN    10.0  
1                        50  25-Jul-80        NaN     9.5  
2                        45  01-Mar-73        NaN     9.0  
3                        44  17-Nov-92          Y     8.5  
4                        43  21-Oct-77        NaN     8.0

Trong bảng, thông tin mỗi bộ phim nằm trong một hàng và các cột:

<ul>
    <li><b>artist</b>: Tên nghệ sĩ</li>
    <li><b>album</b>: Tên album</li>
    <li><b>released_year</b>: Năm phát hành album</li>
    <li><b>length_min_sec</b>: Độ dài album (giờ, phút, giây)</li>
    <li><b>genre</b>: Thể loại album</li>
    <li><b>music_recording_sales_millions</b>: Doanh thu đĩa nhạc (triệu USD) trên <a href="http://www.song-database.com/">[SONG://DATABASE]</a></li>
    <li><b>claimed_sales_millions</b>: Doanh số đã tuyên bố của album (triệu USD) trên <a href="http://www.song-database.com/">[SONG://DATABASE]</a></li>
    <li><b>date_released</b>: Ngày phát hành album</li>
    <li><b>soundtrack</b>: Cho biết album có phải là movie soundtrack (Y - Có) hay không (N - Không)</li>
    <li><b>rating_of_friends</b>: Cho biết xếp hạng từ bạn bè từ 1 tới 10</li>
</ul>

Có thể xem tập dữ liệu bên dưới:

<font size="1">
<table font-size:xx-small style="width:25%">
  <tr>
    <th>Artist</th>
    <th>Album</th> 
    <th>Released</th>
    <th>Length</th>
    <th>Genre</th> 
    <th>Music recording sales (millions)</th>
    <th>Claimed sales (millions)</th>
    <th>Released</th>
    <th>Soundtrack</th>
    <th>Rating (friends)</th>
  </tr>
  <tr>
    <td>Michael Jackson</td>
    <td>Thriller</td> 
    <td>1982</td>
    <td>00:42:19</td>
    <td>Pop, rock, R&B</td>
    <td>46</td>
    <td>65</td>
    <td>30-Nov-82</td>
    <td></td>
    <td>10.0</td>
  </tr>
  <tr>
    <td>AC/DC</td>
    <td>Back in Black</td> 
    <td>1980</td>
    <td>00:42:11</td>
    <td>Hard rock</td>
    <td>26.1</td>
    <td>50</td>
    <td>25-Jul-80</td>
    <td></td>
    <td>8.5</td>
  </tr>
    <tr>
    <td>Pink Floyd</td>
    <td>The Dark Side of the Moon</td> 
    <td>1973</td>
    <td>00:42:49</td>
    <td>Progressive rock</td>
    <td>24.2</td>
    <td>45</td>
    <td>01-Mar-73</td>
    <td></td>
    <td>9.5</td>
  </tr>
    <tr>
    <td>Whitney Houston</td>
    <td>The Bodyguard</td> 
    <td>1992</td>
    <td>00:57:44</td>
    <td>Soundtrack/R&B, soul, pop</td>
    <td>26.1</td>
    <td>50</td>
    <td>25-Jul-80</td>
    <td>Y</td>
    <td>7.0</td>
  </tr>
    <tr>
    <td>Meat Loaf</td>
    <td>Bat Out of Hell</td> 
    <td>1977</td>
    <td>00:46:33</td>
    <td>Hard rock, progressive rock</td>
    <td>20.6</td>
    <td>43</td>
    <td>21-Oct-77</td>
    <td></td>
    <td>7.0</td>
  </tr>
    <tr>
    <td>Eagles</td>
    <td>Their Greatest Hits (1971-1975)</td> 
    <td>1976</td>
    <td>00:43:08</td>
    <td>Rock, soft rock, folk rock</td>
    <td>32.2</td>
    <td>42</td>
    <td>17-Feb-76</td>
    <td></td>
    <td>9.5</td>
  </tr>
    <tr>
    <td>Bee Gees</td>
    <td>Saturday Night Fever</td> 
    <td>1977</td>
    <td>1:15:54</td>
    <td>Disco</td>
    <td>20.6</td>
    <td>40</td>
    <td>15-Nov-77</td>
    <td>Y</td>
    <td>9.0</td>
  </tr>
    <tr>
    <td>Fleetwood Mac</td>
    <td>Rumours</td> 
    <td>1977</td>
    <td>00:40:01</td>
    <td>Soft rock</td>
    <td>27.9</td>
    <td>40</td>
    <td>04-Feb-77</td>
    <td></td>
    <td>9.5</td>
  </tr>
</table></font>

In [32]:
x = df[['Length']]
x

Length
0  0:42:19
1  0:42:11
2  0:42:49
3  0:57:44
4  0:46:33
5  0:43:08
6  1:15:54
7  0:40:01

In [33]:
type(x)

pandas.core.frame.DataFrame

In [34]:
y = df[['Artist','Length','Genre']]
y

Artist   Length                        Genre
0  Michael Jackson  0:42:19               pop, rock, R&B
1            AC/DC  0:42:11                    hard rock
2       Pink Floyd  0:42:49             progressive rock
3  Whitney Houston  0:57:44               R&B, soul, pop
4        Meat Loaf  0:46:33  hard rock, progressive rock
5           Eagles  0:43:08   rock, soft rock, folk rock
6         Bee Gees  1:15:54                        disco
7    Fleetwood Mac  0:40:01                    soft rock

In [35]:
df.iloc[0,0]

'Michael Jackson'

In [36]:
df.loc[0, 'Artist']

'Michael Jackson'

In [37]:
df.iloc[0:2, 0:3]

Artist          Album  Released
0  Michael Jackson       Thriller      1982
1            AC/DC  Back in Black      1980

In [38]:
df.head()

Artist                      Album  Released   Length  \
0  Michael Jackson                   Thriller      1982  0:42:19   
1            AC/DC              Back in Black      1980  0:42:11   
2       Pink Floyd  The Dark Side of the Moon      1973  0:42:49   
3  Whitney Houston              The Bodyguard      1992  0:57:44   
4        Meat Loaf            Bat Out of Hell      1977  0:46:33   

                         Genre  Music Recording Sales (millions)  \
0               pop, rock, R&B                              46.0   
1                    hard rock                              26.1   
2             progressive rock                              24.2   
3               R&B, soul, pop                              27.4   
4  hard rock, progressive rock                              20.6   

   Claimed Sales (millions) Released.1 Soundtrack  Rating  
0                        65  30-Nov-82        NaN    10.0  
1                        50  25-Jul-80        NaN     9.5  
2                        45  01-Mar-73        NaN     9.0  
3                        44  17-Nov-92          Y     8.5  
4                        43  21-Oct-77        NaN     8.0

In [39]:
df.loc[0:2, 'Artist':'Genre']

Artist                      Album  Released   Length  \
0  Michael Jackson                   Thriller      1982  0:42:19   
1            AC/DC              Back in Black      1980  0:42:11   
2       Pink Floyd  The Dark Side of the Moon      1973  0:42:49   

              Genre  
0    pop, rock, R&B  
1         hard rock  
2  progressive rock

### **QUIZ**

Sử dụng biến <code> q</code> để lưu cột <b>Rating</b> dưới dạng dataframe

In [40]:
q = df[['Rating']]
q

Rating
0    10.0
1     9.5
2     9.0
3     8.5
4     8.0
5     7.5
6     7.0
7     6.5

Gán biến <code> q</code> cho dataframe được tạo thành từ cột <b>Released</b> và <b>Artist</b>:

In [41]:
q = df[['Released','Artist']]
q

Released           Artist
0      1982  Michael Jackson
1      1980            AC/DC
2      1973       Pink Floyd
3      1992  Whitney Houston
4      1977        Meat Loaf
5      1976           Eagles
6      1977         Bee Gees
7      1977    Fleetwood Mac

Truy cập hàng thứ 2 và cột thứ 3 của <code> df</code>:

In [42]:
df.iloc[1,2]

1980